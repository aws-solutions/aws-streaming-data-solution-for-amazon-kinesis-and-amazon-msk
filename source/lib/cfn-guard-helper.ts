/** ************************************************************************************************
*   Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                             *
*   SPDX-License-Identifier: Apache-2.0                                                            *
 ************************************************************************************************ */

import * as cdk  from 'aws-cdk-lib';

export class CfnGuardHelper {
    public static addSuppressions(resource: cdk.CfnResource, suppressions: string | string[]) {
        let rules = [];

        if (suppressions instanceof Array) {
            for (const suppression of suppressions) {
                rules.push(suppression);
            }
        } else {
            rules.push(suppressions);
        }

        // Ensure that the metadata property exists.
        if (!resource.cfnOptions.metadata) {
            resource.cfnOptions.metadata = {};
        }

        if (resource.cfnOptions.metadata.guard?.SuppressedRules) {
            // If the CfnResource already contains some suppressions, we don't want to erase them.
            const existingRules = resource.cfnOptions.metadata.guard.SuppressedRules;
            rules = [...existingRules, ...rules];
        }

        // It's possible that multiple constructs try to add the same suppression.
        // We only keep one occurrence (last) of each.
        const uniqueRules = [
            ...new Map(
                rules.map(rule => [rule, rule])
            ).values()
        ];

        resource.cfnOptions.metadata.guard = {
            SuppressedRules: uniqueRules
        }
    }
}
