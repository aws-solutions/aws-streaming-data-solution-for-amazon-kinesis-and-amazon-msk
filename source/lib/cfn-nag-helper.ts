/** ************************************************************************************************
*   Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                             *
*   SPDX-License-Identifier: Apache-2.0                                                            *
 ************************************************************************************************ */

import * as cdk  from 'aws-cdk-lib';

export interface CfnNagSuppression {
    Id: string;
    Reason: string;
}

export class CfnNagHelper {
    public static addSuppressions(resource: cdk.CfnResource, suppressions: CfnNagSuppression | CfnNagSuppression[]) {
        let rules = [];

        if (suppressions instanceof Array) {
            for (const suppression of suppressions) {
                rules.push({ id: suppression.Id, reason: suppression.Reason });
            }
        } else {
            rules.push({ id: suppressions.Id, reason: suppressions.Reason });
        }

        if (resource.cfnOptions.metadata?.cfn_nag) {
            // If the CfnResource already contains some suppressions, we don't want to erase them.
            const existingRules = resource.cfnOptions.metadata.cfn_nag.rules_to_suppress;
            rules = [...existingRules, ...rules];
        }

        // It's possible that multiple constructs try to add the same suppression.
        // We only keep one occurrence (last) of each.
        // Based on https://stackoverflow.com/a/56768137
        const uniqueRules = [
            ...new Map(
                rules.map(rule => [rule.id, rule])
            ).values()
        ];

        if (!resource.cfnOptions.metadata) {
            resource.cfnOptions.metadata = {};
        }

        resource.cfnOptions.metadata.cfn_nag = {
            rules_to_suppress: uniqueRules
        }
    }
}
