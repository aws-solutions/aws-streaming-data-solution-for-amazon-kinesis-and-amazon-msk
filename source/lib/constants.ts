/*********************************************************************************************************************
 *  Copyright 2020-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.                                      *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

export default class Constants {
    private static readonly ssmManagedActions = [
        'ssm:DescribeAssociation',
        'ssm:GetDeployablePatchSnapshotForInstance',
        'ssm:GetDocument',
        'ssm:DescribeDocument',
        'ssm:GetManifest',
        'ssm:GetParameter',
        'ssm:GetParameters',
        'ssm:ListAssociations',
        'ssm:ListInstanceAssociations',
        'ssm:PutInventory',
        'ssm:PutComplianceItems',
        'ssm:PutConfigurePackageResult',
        'ssm:UpdateAssociationStatus',
        'ssm:UpdateInstanceAssociationStatus',
        'ssm:UpdateInstanceInformation',
        'ec2messages:AcknowledgeMessage',
        'ec2messages:DeleteMessage',
        'ec2messages:FailMessage',
        'ec2messages:GetEndpoint',
        'ec2messages:GetMessages',
        'ec2messages:SendReply',
        'ssmmessages:CreateControlChannel',
        'ssmmessages:CreateDataChannel',
        'ssmmessages:OpenControlChannel',
        'ssmmessages:OpenDataChannel'
    ]

    public static get SsmManagedActions(): string[] {
        return Constants.ssmManagedActions;
    }
}