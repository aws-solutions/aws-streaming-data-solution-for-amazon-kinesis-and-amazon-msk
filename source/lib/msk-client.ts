/*********************************************************************************************************************
 *  Copyright 2020-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.                                      *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import * as cdk from '@aws-cdk/core';
import * as ec2 from '@aws-cdk/aws-ec2';
import * as iam from '@aws-cdk/aws-iam';

export interface KafkaClientProps {
    readonly vpcId: string;
    readonly subnetId: string;
    readonly imageId: string;
    readonly instanceType: string;

    readonly kafkaVersion: string;
    readonly clusterSecurityGroupId: string;
}

export class KafkaClient extends cdk.Construct {
    private readonly Instance: ec2.CfnInstance;

    public get InstanceId(): string {
        return this.Instance.ref;
    }

    constructor(scope: cdk.Construct, id: string, props: KafkaClientProps) {
        super(scope, id);

        const instanceProfile = this.createInstanceProfile();

        const userDataCommands = [
            '#!/bin/bash',
            'yum update -y',
            'yum install java-1.8.0 python3 -y',

            'mkdir -p /home/kafka && cd /home/kafka',
            `wget https://archive.apache.org/dist/kafka/${props.kafkaVersion}/kafka_2.12-${props.kafkaVersion}.tgz`,
            `tar -xzf kafka_2.12-${props.kafkaVersion}.tgz --strip 1 && rm kafka_2.12-${props.kafkaVersion}.tgz`,

            `find /usr/lib/jvm/ -name "cacerts" | xargs -I '{}' cp '{}' /tmp/kafka.client.truststore.jks`,
            `touch bin/client.properties`,
            `echo "security.protocol=SSL" >> bin/client.properties`,
            `echo "ssl.truststore.location=/tmp/kafka.client.truststore.jks" >> bin/client.properties`,
        ];

        this.Instance = new ec2.CfnInstance(this, 'Client', {
            imageId: props.imageId,
            instanceType: props.instanceType,
            subnetId: props.subnetId,
            iamInstanceProfile: instanceProfile.ref,
            securityGroupIds: [props.clusterSecurityGroupId],
            userData: cdk.Fn.base64(userDataCommands.join('\n')),
            tags: [{ key: 'Name', value: 'KafkaClient' }],
        })
    }

    private createInstanceProfile(): iam.CfnInstanceProfile {
        const role = new iam.Role(this, 'Role', {
            assumedBy: new iam.ServicePrincipal('ec2.amazonaws.com')
        });

        const ssmPolicy = new iam.Policy(this, 'SessionManagerPolicy', {
            statements: [new iam.PolicyStatement({
                resources: ['*'],
                actions: [
                    'ssm:UpdateInstanceInformation',
                    'ssmmessages:CreateControlChannel',
                    'ssmmessages:CreateDataChannel',
                    'ssmmessages:OpenControlChannel',
                    'ssmmessages:OpenDataChannel'
                ]
            })]
        });

        this.addW12Suppression(ssmPolicy, 'Session Manager actions do not support resource level permissions');
        ssmPolicy.attachToRole(role);

        const mskPolicy = new iam.Policy(this, 'MskPolicy', {
            statements: [new iam.PolicyStatement({
                resources: ['*'],
                actions: ['kafka:DescribeCluster', 'kafka:GetBootstrapBrokers']
            })]
        });

        this.addW12Suppression(mskPolicy, 'MSK actions do not support resource level permissions');
        mskPolicy.attachToRole(role);

        const cfnRole = role.node.defaultChild as iam.CfnRole;
        return new iam.CfnInstanceProfile(this, 'InstanceProfile', {
            roles: [cfnRole.ref]
        });
    }

    private addW12Suppression(policy: iam.Policy, reason: string) {
        const cfnPolicy = policy.node.defaultChild as iam.CfnPolicy;

        cfnPolicy.cfnOptions.metadata = {
            cfn_nag: {
                rules_to_suppress: [{
                    id: 'W12',
                    reason: reason
                }]
            }
        };
    }
}
