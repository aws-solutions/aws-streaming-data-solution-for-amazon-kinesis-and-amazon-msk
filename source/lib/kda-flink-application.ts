/** ************************************************************************************************
*   Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                             *
*   SPDX-License-Identifier: Apache-2.0                                                            *
 ************************************************************************************************ */

import * as cdk  from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { aws_kinesisanalytics as analytics, aws_iam as iam } from 'aws-cdk-lib';

import { CfnNagHelper } from './cfn-nag-helper';
import { CfnGuardHelper } from './cfn-guard-helper';
import { FlinkBase, FlinkBaseProps } from './kda-base';

export interface FlinkApplicationProps extends FlinkBaseProps {
    readonly environmentProperties: analytics.CfnApplicationV2.PropertyGroupProperty;
    readonly metricsLevel: string;

    readonly codeBucketArn: string;
    readonly codeFileKey: string;

    readonly enableSnapshots: string;
    readonly enableAutoScaling: string;
}

export class FlinkApplication extends FlinkBase {
    constructor(scope: Construct, id: string, props: FlinkApplicationProps) {
        super(scope, id, props);
        this.addCfnNagSuppressions();
        this.addCfnGuardSuppressions();
    }

    protected createApplication(props: FlinkApplicationProps): analytics.CfnApplicationV2 {
        const autoScalingCondition = new cdk.CfnCondition(this, 'EnableAutoScaling', {
            expression: cdk.Fn.conditionEquals(props.enableAutoScaling, 'true')
        });

        const snapshotCondition = new cdk.CfnCondition(this, 'EnableSnapshots', {
            expression: cdk.Fn.conditionEquals(props.enableSnapshots, 'true')
        });

        return new analytics.CfnApplicationV2(this, 'Application', {
            runtimeEnvironment: 'FLINK-1_15',
            serviceExecutionRole: this.Role.roleArn,
            applicationConfiguration: {
                applicationCodeConfiguration: {
                    codeContent: {
                        s3ContentLocation: {
                            bucketArn: props.codeBucketArn,
                            fileKey: props.codeFileKey
                        }
                    },
                    codeContentType: 'ZIPFILE'
                },
                environmentProperties: {
                    propertyGroups: [props.environmentProperties]
                },
                flinkApplicationConfiguration: {
                    monitoringConfiguration: {
                        configurationType: 'CUSTOM',
                        logLevel: props.logLevel,
                        metricsLevel: props.metricsLevel
                    },
                    parallelismConfiguration: {
                        configurationType: 'CUSTOM',
                        autoScalingEnabled: cdk.Fn.conditionIf(autoScalingCondition.logicalId, true, false)
                    },
                    checkpointConfiguration: {
                        configurationType: 'DEFAULT'
                    }
                },
                applicationSnapshotConfiguration: {
                    snapshotsEnabled: cdk.Fn.conditionIf(snapshotCondition.logicalId, true, false)
                }
            }
        });
    }

    protected createRole(props: FlinkApplicationProps): iam.IRole {
        const s3CodePolicy = new iam.PolicyDocument({
            statements: [new iam.PolicyStatement({
                resources: [`${props.codeBucketArn}/${props.codeFileKey}`],
                actions: ['s3:GetObjectVersion', 's3:GetObject']
            })]
        });

        const logsPolicy = new iam.PolicyDocument({
            statements: [
                new iam.PolicyStatement({
                    resources: [`arn:${cdk.Aws.PARTITION}:logs:${cdk.Aws.REGION}:${cdk.Aws.ACCOUNT_ID}:log-group:*`],
                    actions: ['logs:DescribeLogGroups']
                }),
                new iam.PolicyStatement({
                    resources: [this.LogGroup.logGroupArn],
                    actions: ['logs:DescribeLogStreams', 'logs:PutLogEvents']
                })
            ]
        });

        const vpcPolicy = new iam.PolicyDocument({
            statements: [
                new iam.PolicyStatement({
                    resources: ['*'],
                    actions: [
                        'ec2:CreateNetworkInterface',
                        'ec2:DescribeNetworkInterfaces',
                        'ec2:DescribeVpcs',
                        'ec2:DeleteNetworkInterface',
                        'ec2:DescribeDhcpOptions',
                        'ec2:DescribeSubnets',
                        'ec2:DescribeSecurityGroups'
                    ]
                }),
                new iam.PolicyStatement({
                    resources: [`arn:${cdk.Aws.PARTITION}:ec2:${cdk.Aws.REGION}:${cdk.Aws.ACCOUNT_ID}:network-interface/*`],
                    actions: ['ec2:CreateNetworkInterfacePermission']
                })
            ]
        });

        return new iam.Role(this, 'AppRole', {
            assumedBy: new iam.ServicePrincipal('kinesisanalytics.amazonaws.com'),
            inlinePolicies: {
                S3Policy: s3CodePolicy,
                LogsPolicy: logsPolicy,
                VpcPolicy: vpcPolicy
            }
        });
    }

    private addCfnNagSuppressions() {
        const cfnRole = this.Role.node.defaultChild as iam.CfnRole;
        CfnNagHelper.addSuppressions(cfnRole, {
            Id: 'W11',
            Reason: 'EC2 actions in VPC policy do not support resource level permissions'
        });
    }

    private addCfnGuardSuppressions() {
        const cfnRole = this.Role.node.defaultChild as iam.CfnRole;
        CfnGuardHelper.addSuppressions(cfnRole, ['IAM_NO_INLINE_POLICY_CHECK']);
    }
}
