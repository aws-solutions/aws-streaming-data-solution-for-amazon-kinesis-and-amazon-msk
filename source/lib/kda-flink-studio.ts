/*********************************************************************************************************************
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                                                *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import * as cdk from '@aws-cdk/core';
import * as analytics from '@aws-cdk/aws-kinesisanalytics';
import * as iam from '@aws-cdk/aws-iam';
import * as glue from '@aws-cdk/aws-glue';

import { CfnNagHelper } from './cfn-nag-helper';
import { FlinkBase, FlinkBaseProps } from './kda-base';

export class FlinkStudio extends FlinkBase {
    private DatabaseName: string = '';

    constructor(scope: cdk.Construct, id: string, props: FlinkBaseProps) {
        super(scope, id, props);
        this.addCfnNagSuppressions();
    }

    protected createRole(_props: FlinkBaseProps): iam.IRole {
        const glueDb = new glue.CfnDatabase(this, 'Database', {
            catalogId: cdk.Aws.ACCOUNT_ID,
            databaseInput: {
                description: `${cdk.Aws.STACK_NAME} - Database for Amazon Kinesis Data Analytics Studio`
            }
        });

        this.DatabaseName = glueDb.ref;

        const logsPolicy = new iam.PolicyDocument({
            statements: [
                new iam.PolicyStatement({
                    resources: [`arn:${cdk.Aws.PARTITION}:logs:${cdk.Aws.REGION}:${cdk.Aws.ACCOUNT_ID}:log-group:*`],
                    actions: ['logs:DescribeLogGroups']
                }),
                new iam.PolicyStatement({
                    resources: [this.LogGroup.logGroupArn],
                    actions: ['logs:DescribeLogStreams', 'logs:PutLogEvents']
                })
            ]
        });

        const vpcPolicy = new iam.PolicyDocument({
            statements: [
                new iam.PolicyStatement({
                    resources: ['*'],
                    actions: [
                        'ec2:CreateNetworkInterface',
                        'ec2:DescribeNetworkInterfaces',
                        'ec2:DescribeVpcs',
                        'ec2:DeleteNetworkInterface',
                        'ec2:DescribeDhcpOptions',
                        'ec2:DescribeSubnets',
                        'ec2:DescribeSecurityGroups'
                    ]
                }),
                new iam.PolicyStatement({
                    resources: [`arn:${cdk.Aws.PARTITION}:ec2:${cdk.Aws.REGION}:${cdk.Aws.ACCOUNT_ID}:network-interface/*`],
                    actions: ['ec2:CreateNetworkInterfacePermission']
                })
            ]
        });

        // This policy is based on https://docs.aws.amazon.com/kinesisanalytics/latest/java/how-zeppelin-appendix-iam.html
        const gluePolicy = new iam.PolicyDocument({
            statements: [
                new iam.PolicyStatement({
                    resources: [
                        `arn:${cdk.Aws.PARTITION}:glue:${cdk.Aws.REGION}:${cdk.Aws.ACCOUNT_ID}:connection/*`,
                        `arn:${cdk.Aws.PARTITION}:glue:${cdk.Aws.REGION}:${cdk.Aws.ACCOUNT_ID}:table/${this.DatabaseName}/*`,
                        `arn:${cdk.Aws.PARTITION}:glue:${cdk.Aws.REGION}:${cdk.Aws.ACCOUNT_ID}:database/${this.DatabaseName}`,
                        `arn:${cdk.Aws.PARTITION}:glue:${cdk.Aws.REGION}:${cdk.Aws.ACCOUNT_ID}:database/hive`,
                        `arn:${cdk.Aws.PARTITION}:glue:${cdk.Aws.REGION}:${cdk.Aws.ACCOUNT_ID}:catalog`,
                        `arn:${cdk.Aws.PARTITION}:glue:${cdk.Aws.REGION}:${cdk.Aws.ACCOUNT_ID}:userDefinedFunction/*`
                    ],
                    actions: [
                        'glue:GetConnection',
                        'glue:GetTable',
                        'glue:GetTables',
                        'glue:GetDatabase',
                        'glue:CreateTable',
                        'glue:UpdateTable',
                        'glue:GetUserDefinedFunction'
                    ]
                }),
                new iam.PolicyStatement({
                    resources: ['*'],
                    actions: ['glue:GetDatabases']
                }),
            ]
        });

        return new iam.Role(this, 'AppRole', {
            assumedBy: new iam.ServicePrincipal('kinesisanalytics.amazonaws.com'),
            inlinePolicies: {
                LogsPolicy: logsPolicy,
                VpcPolicy: vpcPolicy,
                GluePolicy: gluePolicy
            }
        });
    }

    protected createApplication(_props: FlinkBaseProps): analytics.CfnApplicationV2 {
        return new analytics.CfnApplicationV2(this, 'Studio', {
            runtimeEnvironment: 'ZEPPELIN-FLINK-2_0',
            applicationMode: 'INTERACTIVE',
            serviceExecutionRole: this.Role.roleArn,
            applicationConfiguration: {
                zeppelinApplicationConfiguration: {
                    catalogConfiguration: {
                        glueDataCatalogConfiguration: {
                            databaseArn: `arn:${cdk.Aws.PARTITION}:glue:${cdk.Aws.REGION}:${cdk.Aws.ACCOUNT_ID}:database/${this.DatabaseName}`
                        }
                    },
                    monitoringConfiguration: {
                        logLevel: _props.logLevel
                    },
                    customArtifactsConfiguration: [
                        {
                            artifactType: 'DEPENDENCY_JAR',
                            mavenReference: {
                                groupId: 'org.apache.flink',
                                artifactId: 'flink-sql-connector-kinesis_2.12',
                                version: '1.13.2'
                            }
                        },
                        {
                            artifactType: 'DEPENDENCY_JAR',
                            mavenReference: {
                                groupId: 'org.apache.flink',
                                artifactId: 'flink-connector-kafka_2.12',
                                version: '1.13.2'
                            }
                        },
                        {
                            artifactType: 'DEPENDENCY_JAR',
                            mavenReference: {
                                groupId: 'software.amazon.msk',
                                artifactId: 'aws-msk-iam-auth',
                                version: '1.1.0'
                            }
                        }
                    ]
                }
            }
        });
    }

    private addCfnNagSuppressions() {
        const cfnRole = this.Role.node.defaultChild as iam.CfnRole;
        CfnNagHelper.addSuppressions(cfnRole, {
            Id: 'W11',
            Reason: 'EC2 actions do not support resource level permissions / Studio uses default Glue database'
        });
    }
}
