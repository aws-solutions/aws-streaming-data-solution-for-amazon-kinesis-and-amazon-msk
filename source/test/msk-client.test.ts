/*********************************************************************************************************************
 *  Copyright 2020-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.                                      *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import * as cdk from '@aws-cdk/core';
import { SynthUtils, expect as expectCDK, haveResource, ResourcePart } from '@aws-cdk/assert';

import { KafkaClient } from '../lib/msk-client';

let stack: cdk.Stack;

beforeEach(() => {
    const app = new cdk.App();
    stack = new cdk.Stack(app, 'TestStack');
});

test('creates a Kafka client instance', () => {
    const ec2Client = new KafkaClient(stack, 'TestClient', {
        vpcId: 'vpc-123',
        subnetId: 'subnet-abc',
        imageId: 'ami-1234',
        instanceType: 't3.small',

        clusterSecurityGroupId: 'cluster-sg',
        clusterName: 'my-cluster',
        kafkaVersion: '2.6.0'
    });

    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
    expect(ec2Client.InstanceId).not.toBeUndefined();
});

test('adds cfn_nag suppressions', () => {
    new KafkaClient(stack, 'TestClient', {
        vpcId: 'vpc-123',
        subnetId: 'subnet-abc',
        imageId: 'ami-1234',
        instanceType: 't3.small',

        clusterSecurityGroupId: 'cluster-sg',
        clusterName: 'my-cluster',
        kafkaVersion: '2.6.0'
    });

    expectCDK(stack).to(haveResource('AWS::IAM::Policy', {
        Metadata: {
            cfn_nag: {
                rules_to_suppress: [{
                    id: 'W12',
                    reason: 'Session Manager actions do not support resource level permissions'
                }]
            }
        }
    }, ResourcePart.CompleteDefinition));

    expectCDK(stack).to(haveResource('AWS::IAM::Policy', {
        Metadata: {
            cfn_nag: {
                rules_to_suppress: [{
                    id: 'W12',
                    reason: 'MSK actions do not support resource level permissions'
                }]
            }
        }
    }, ResourcePart.CompleteDefinition));
});
