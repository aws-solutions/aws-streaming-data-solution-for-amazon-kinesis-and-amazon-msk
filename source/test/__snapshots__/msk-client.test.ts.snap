// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates a Kafka client instance 1`] = `
Object {
  "Resources": Object {
    "TestClientD3610C34": Object {
      "Properties": Object {
        "IamInstanceProfile": Object {
          "Ref": "TestClientInstanceProfileF9C60143",
        },
        "ImageId": "ami-1234",
        "InstanceType": "t3.small",
        "SecurityGroupIds": Array [
          "cluster-sg",
        ],
        "SubnetId": "subnet-abc",
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "KafkaClient",
          },
        ],
        "UserData": Object {
          "Fn::Base64": "#!/bin/bash
yum update -y
yum install java-11-amazon-corretto-headless python3 -y
mkdir -p /home/kafka && cd /home/kafka
wget https://archive.apache.org/dist/kafka/2.6.0/kafka_2.12-2.6.0.tgz
tar -xzf kafka_2.12-2.6.0.tgz --strip 1 && rm kafka_2.12-2.6.0.tgz
wget https://github.com/aws/aws-msk-iam-auth/releases/download/v1.1.1/aws-msk-iam-auth-1.1.1-all.jar
wget https://github.com/aws/aws-msk-iam-auth/releases/download/v1.1.1/aws-msk-iam-auth-1.1.1-all.jar.sha256
IAM_LIB_CHECKSUM=\`cat aws-msk-iam-auth-1.1.1-all.jar.sha256\`
echo \\"$IAM_LIB_CHECKSUM aws-msk-iam-auth-1.1.1-all.jar\\" | sha256sum -c
mv aws-msk-iam-auth-1.1.1-all.jar aws-msk-iam-auth-1.1.1-all.jar.sha256 ./libs
find /usr/lib/jvm/ -name \\"cacerts\\" | xargs -I '{}' cp '{}' /tmp/kafka.client.truststore.jks
touch bin/client-ssl.properties
echo \\"security.protocol=SSL\\" >> bin/client-ssl.properties
echo \\"ssl.truststore.location=/tmp/kafka.client.truststore.jks\\" >> bin/client-ssl.properties
touch bin/client-sasl.properties
echo \\"security.protocol=SASL_SSL\\" >> bin/client-sasl.properties
echo \\"sasl.mechanism=SCRAM-SHA-512\\" >> bin/client-sasl.properties
echo \\"ssl.truststore.location=/tmp/kafka.client.truststore.jks\\" >> bin/client-sasl.properties
touch bin/client-iam.properties
echo \\"security.protocol=SASL_SSL\\" >> bin/client-iam.properties
echo \\"sasl.mechanism=AWS_MSK_IAM\\" >> bin/client-iam.properties
echo \\"sasl.jaas.config=software.amazon.msk.auth.iam.IAMLoginModule required;\\" >> bin/client-iam.properties
echo \\"sasl.client.callback.handler.class=software.amazon.msk.auth.iam.IAMClientCallbackHandler\\" >> bin/client-iam.properties",
        },
      },
      "Type": "AWS::EC2::Instance",
    },
    "TestClientInstanceProfileF9C60143": Object {
      "Properties": Object {
        "Roles": Array [
          Object {
            "Ref": "TestClientRole00F4BDF5",
          },
        ],
      },
      "Type": "AWS::IAM::InstanceProfile",
    },
    "TestClientMskPolicyDC57001A": Object {
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W12",
              "reason": "MSK actions do not support resource level permissions",
            },
          ],
        },
      },
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "kafka:DescribeCluster",
                "kafka:GetBootstrapBrokers",
              ],
              "Effect": "Allow",
              "Resource": "*",
              "Sid": "ClusterMetadata",
            },
            Object {
              "Action": Array [
                "kafka-cluster:Connect",
                "kafka-cluster:DescribeCluster",
                "kafka-cluster:AlterGroup",
                "kafka-cluster:DescribeGroup",
                "kafka-cluster:DeleteGroup",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":kafka:",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":cluster/my-cluster/*",
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":kafka:",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":group/my-cluster/*/*",
                    ],
                  ],
                },
              ],
              "Sid": "ClusterAPIs",
            },
            Object {
              "Action": Array [
                "kafka-cluster:*Topic*",
                "kafka-cluster:WriteData",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":kafka:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":topic/my-cluster/*/*",
                  ],
                ],
              },
              "Sid": "ProducerAPIs",
            },
            Object {
              "Action": Array [
                "kafka-cluster:*Topic*",
                "kafka-cluster:ReadData",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":kafka:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":topic/my-cluster/*/*",
                  ],
                ],
              },
              "Sid": "ConsumerAPIs",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TestClientMskPolicyDC57001A",
        "Roles": Array [
          Object {
            "Ref": "TestClientRole00F4BDF5",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TestClientRole00F4BDF5": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "ec2.",
                      Object {
                        "Ref": "AWS::URLSuffix",
                      },
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "TestClientSessionManagerPolicy40AA67DF": Object {
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W12",
              "reason": "Session Manager actions do not support resource level permissions",
            },
          ],
        },
      },
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "ssm:UpdateInstanceInformation",
                "ssmmessages:CreateControlChannel",
                "ssmmessages:CreateDataChannel",
                "ssmmessages:OpenControlChannel",
                "ssmmessages:OpenDataChannel",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TestClientSessionManagerPolicy40AA67DF",
        "Roles": Array [
          Object {
            "Ref": "TestClientRole00F4BDF5",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
